File: backend/app/schemas/certificate.py
"""
Certificate schemas for API requests and responses.
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class CertificateBase(BaseModel):
    """Base certificate schema."""
    title: str
    description: Optional[str] = None


class CertificateCreate(CertificateBase):
    """Schema for creating certificates."""
    user_id: int
    course_id: int


class CertificateResponse(CertificateBase):
    """Schema for certificate responses."""
    id: int
    user_id: int
    course_id: int
    certificate_id: str
    verification_code: str
    certificate_url: Optional[str] = None
    is_verified: bool
    issued_at: datetime
    expires_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class CertificateWithDetails(CertificateResponse):
    """Certificate response with user and course details."""
    user_name: str
    course_title: str
    instructor_name: str

    class Config:
        from_attributes = True


class CompletionStatus(BaseModel):
    """Schema for course completion status."""
    completed: bool
    lectures_completed: bool
    quizzes_passed: bool
    lecture_progress: dict
    quiz_details: list
    completion_percentage: float
    error: Optional[str] = None


class CertificateVerification(BaseModel):
    """Schema for certificate verification."""
    valid: bool
    certificate: Optional[CertificateWithDetails] = None
    message: str

File: frontend/app/types/certificate.ts
/**
 * Certificate-related type definitions
 */

export interface Certificate {
    id: number;
    user_id: number;
    course_id: number;
    certificate_id: string;
    title: string;
    description?: string;
    verification_code: string;
    certificate_url?: string;
    is_verified: boolean;
    issued_at: string;
    expires_at?: string;
}

export interface CertificateWithDetails extends Certificate {
    user_name: string;
    course_title: string;
    instructor_name: string;
}

export interface CompletionStatus {
    completed: boolean;
    lectures_completed: boolean;
    quizzes_passed: boolean;
    lecture_progress: {
        completed: number;
        total: number;
    };
    quiz_details: Array<{
        quiz_id: number;
        title: string;
        passed: boolean;
        score?: number;
        passing_score: number;
    }>;
    completion_percentage: number;
    error?: string;
}

export interface CertificateVerification {
    valid: boolean;
    certificate?: CertificateWithDetails;
    message: string;
}

export interface CertificateCreateRequest {
    user_id: number;
    course_id: number;
}